pkg/plugins/builtin/script/script.go

```diff
@@ -0,0 +1,147 @@
+package script
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	gexec "os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/ovh/utask"
+
+	"github.com/ovh/utask/pkg/plugins/taskplugin"
+)
+
+// the script plugin execute scripts
+var (
+	Plugin = taskplugin.New("script", "0.1", exec,
+		taskplugin.WithConfig(validConfig, Config{}),
+	)
+)
+
+// Metadata represents the metadata of script execution
+type Metadata struct {
+	ExitCode      string `json:"exit_code"`
+	ProcessState  string `json:"process_state"`
+	Output        string `json:"output"`
+	ExecutionTime string `json:"execution_time"`
+	Error         string `json:"error"`
+}
+
+// Config is the configuration needed to execute a script
+type Config struct {
+	File    string   `json:"file"`
+	Argv    []string `json:"argv,omitempty"`
+	Timeout string   `json:"timeout,omitempty"`
+	Stdin   string   `json:"stdin,omitempty"`
+}
+
+func validConfig(config interface{}) error {
+	cfg := config.(*Config)
+
+	if cfg.File == "" {
+		return errors.New("file is missing")
+	}
+
+	f, err := os.Stat(filepath.Join(utask.FScriptsFolder, cfg.File))
+	if os.IsNotExist(err) {
+		return fmt.Errorf("%s not found in FS: %s", cfg.File, err.Error())
+	}
+
+	if f.Mode()&0010 == 0 {
+		return fmt.Errorf("%s haven't exec permissions", cfg.File)
+	}
+
+	if cfg.Timeout != "" {
+		if cfg.Timeout[0] == '-' {
+			return errors.New("timeout must be positive")
+		}
+		if _, err := time.ParseDuration(cfg.Timeout); err != nil {
+			return fmt.Errorf("timeout is wrong %s", err.Error())
+		}
+	}
+
+	return nil
+}
+
+func exec(stepName string, config interface{}, ctx interface{}) (interface{}, interface{}, error) {
+	cfg := config.(*Config)
+
+	var timeout time.Duration
+
+	if cfg.Timeout != "" {
+		timeout, _ = time.ParseDuration(cfg.Timeout)
+	} else {
+		// default is 2 * 1 minute = 2 minutes
+		timeout = time.Duration(2) * time.Minute
+	}
+
+	ctxe, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cmd := gexec.CommandContext(ctxe, filepath.Join(utask.FScriptsFolder, cfg.File), cfg.Argv...)
+	cmd.Dir = utask.FScriptsFolder
+	cmd.Stdin = strings.NewReader(cfg.Stdin)
+
+	exitCode := 0
+	metaError := ""
+
+	// start exec time timer
+	timer := time.Now()
+	// execute script
+	out, err := cmd.CombinedOutput()
```

Comment:
Why do you want to combine Stdout & Stderr?

As you may expect some JSON as script output, it is maybe better to keep them separated because there is no guarantee on how they are combined

Reply:
It's the responsibility of the script's writer to handle this kind of issue.
Otherwise, if you want to debug, as discussed /w @loopfz, you want to keep a consistent trace of the program execution.

Reply:
It can't be the responsability of the script's writer as he can't control how YOU merge the two streams and then could conduct to unconsistant behaviour.

If the script output one line to stdout and one line to stderr, you will either have stdout first and stderr last or the opposite.

So it is better to split the two streams (and fix the ssh plugin if it does the same) or you should document that the script should write either to stdout or to stderr but not both as you can't guarantee how you'll handle it.

Reply:
Actually all the tests I wrote (even those not implemented in go_test) who had two steams (stdout/stderr) keep the consistency between common terminal execution and `CombinedOutput()` func. IMO it do the job.
I'll add documentation about that but for the moment we should keep it like this.
If there is issue in the future, I'll do another PR to fix this plugin and SSH.